src/Microsoft.Health.Fhir.Core/Features/Operations/ConvertData/ContainerRegistryTemplateProvider.cs

```diff
@@ -73,7 +79,31 @@ async Task<string> TokenEntryFactory(ICacheEntry entry)
 
             try
             {
-                var provider = _templateCollectionProviderFactory.CreateTemplateCollectionProvider(request.TemplateCollectionReference, accessToken);
+                /*
+                   First try to get the template provider from the cache.
+                   If template provider is not in cache, then limit the number of threads that can create new providers to 1, and then create a new provider.
+                   This is to limit the amount of providers created in a multi-threaded scenario, so that we do not have multiple providers pulling the same image from the ACR.
```

Comment:
Just curious, I know you mentioned that the ACR eventually starts throwing 500s when there are many requests for the same image. Does this not happen if there are many requests for different images within the same ACR?

Reply:
It seems related to the total number of requests, or total bandwidth pulling from the ACR in general. So it seems to still throw an error even when pulling different images.

[The ACR service tier limits/throttling docs](https://learn.microsoft.com/en-us/azure/container-registry/container-registry-skus) do state that an HTTP 429 should be returned when too many requests are made, but it appears that is not really what is being observed. It could be that the server (ACR) is returning the appropriate error and the client (ACR SDK client) is not handling it for whatever reason. Or that the server is just denying the connection. But in any case we probably should be just making 1 (or maybe a few) requests to pull from the ACR instead of potentially hundreds. I think that was the intent with the caching in the original implementation but think it just needed to be tweaked slightly to account for concurrency.

For reference the error is:

Pull Image Failed.System.Net.Http.HttpRequestException: The SSL connection could not be established, see inner exception.
---> System.IO.IOException:  Received an unexpected EOF or 0 bytes from the transport stream.
at System.Net.Security.SslStream.ReceiveBlobAsync[TIOAdapter](TIOAdapter adapter)
at System.Net.Security.SslStream.ForceAuthenticationAsync[TIOAdapter](TIOAdapter adapter, Boolean receiveFirst, Byte[] reAuthenticationData, Boolean isApm)
at System.Net.Http.ConnectHelper.EstablishSslConnectionAsyncCore(Boolean async, Stream stream, SslClientAuthenticationOptions sslOptions, CancellationToken cancellationToken)
--- End of inner exception stack trace ---
at Microsoft.Rest.RetryDelegatingHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
at System.Net.Http.HttpClient.SendAsyncCore(HttpRequestMessage request, HttpCompletionOption completionOption, Boolean async, Boolean emitTelemetryStartStop, CancellationToken cancellationToken)
at Microsoft.Azure.ContainerRegistry.BlobOperations.GetWithHttpMessagesAsync(String name, String digest, Dictionary`2 customHeaders, CancellationToken cancellationToken)
at Microsoft.Azure.ContainerRegistry.BlobOperationsExtensions.GetAsync(IBlobOperations operations, String name, String digest, CancellationToken cancellationToken)
at Microsoft.Health.Fhir.TemplateManagement.Client.AcrClient.PullBlobAsStreamAcync(String name, String digest, CancellationToken cancellationToken)